crates/wasm_interp/tests/test_i32.rs

```diff
@@ -0,0 +1,280 @@
+    test_u32_binop(op, u32::MAX, u32::MIN, 1);
+}
+
+#[test]
+fn test_i32clz() {
+    let op = I32CLZ;
+    test_i32_unop(op, 0, 32);
+    test_i32_unop(op, -1, 0);
+    test_i32_unop(op, 1, 31);
+    test_i32_unop(op, 1024, 21);
+}
+
+#[test]
+fn test_i32ctz() {
+    let op = I32CTZ;
+    test_i32_unop(op, 0, 32);
+    test_i32_unop(op, -1, 0);
+    test_i32_unop(op, 2, 1);
+    test_i32_unop(op, 1024, 10);
+}
+
+#[test]
+fn test_i32popcnt() {
+    let op = I32POPCNT;
+    test_i32_unop(op, 0, 0);
+    test_i32_unop(op, -1, 32);
+    test_i32_unop(op, 2, 1);
+    test_i32_unop(op, 96, 2);
+}
+
+#[test]
+fn test_i32add() {
+    let op = I32ADD;
+    test_i32_binop(op, 0, 0, 0);
+    test_i32_binop(op, -1, -1, -2);
+    test_i32_binop(op, 1, 1, 2);
+    test_i32_binop(op, i32::MAX, 1, i32::MIN);
+}
+
+#[test]
+fn test_i32sub() {
+    let op = I32SUB;
+    test_i32_binop(op, 0, 0, 0);
+    test_i32_binop(op, -1, 1, -2);
+    test_i32_binop(op, 1, 1, 0);
+    test_i32_binop(op, i32::MIN, 1, i32::MAX);
+}
+
+#[test]
+fn test_i32mul() {
+    let op = I32MUL;
+    test_i32_binop(op, 0, 0, 0);
+    test_i32_binop(op, -1, -1, 1);
+    test_i32_binop(op, 2, 3, 6);
+    test_i32_binop(op, i32::MAX, 2, -2);
+}
+
+#[test]
+fn test_i32divs() {
+    let op = I32DIVS;
+    test_i32_binop(op, -1, -1, 1);
+    test_i32_binop(op, 6, 3, 2);
+    test_i32_binop(op, i32::MIN, -1, i32::MIN);
+}
+
+#[test]
+#[should_panic(expected = "divide by zero")]
+fn test_i32divs_zero() {
+    test_i32_binop(I32DIVS, 1, 0, -1);
+}
+
+#[test]
+fn test_i32divu() {
+    let op = I32DIVU;
+    test_u32_binop(op, 1, 1, 1);
+    test_u32_binop(op, 6, 3, 2);
+}
+
+#[test]
+#[should_panic(expected = "divide by zero")]
+fn test_i32divu_zero() {
+    test_i32_binop(I32DIVU, 1, 0, -1);
+}
+
+#[test]
+fn test_i32rems() {
+    let op = I32REMS;
+    test_i32_binop(op, 5, 2, 1);
+    // test_i32_binop(op, 5, -2, 1); // TODO: we don't match Wasmer, we get 0
+    test_i32_binop(op, -5, 2, -1);
+    // test_i32_binop(op, -5, -2, -1); // TODO: we don't match Wasmer, we get 0
+}
+
+#[test]
+#[should_panic(expected = "divisor of zero")]
+fn test_i32rems_zero() {
+    test_i32_binop(I32REMS, 1, 0, -1);
+}
+
+#[test]
+fn test_i32remu() {
+    let op = I32REMU;
+    test_i32_binop(op, 5, 2, 1);
+}
+
+#[test]
+#[should_panic(expected = "divisor of zero")]
+fn test_i32remu_zero() {
+    test_i32_binop(I32REMU, 1, 0, -1);
+}
+
+#[test]
+fn test_i32and() {
+    test_u32_binop(I32AND, 0x0000_ffff, 0x00ff_00ff, 0x0000_00ff);
+}
+
+#[test]
+fn test_i32or() {
+    test_u32_binop(I32OR, 0x0000_ffff, 0x00ff_00ff, 0x00ff_ffff);
+}
+
+#[test]
+fn test_i32xor() {
+    test_u32_binop(I32XOR, 0x0000_ffff, 0x00ff_00ff, 0x00ff_ff00);
+}
+
+#[test]
+fn test_i32shl() {
+    test_u32_binop(I32SHL, 0xffff_ffff, 8, 0xffff_ff00);
+}
+
+#[test]
+fn test_i32shrs() {
+    test_u32_binop(I32SHRS, 0xffff_0000, 8, 0xffff_ff00);
+}
+
+#[test]
+fn test_i32shru() {
+    test_u32_binop(I32SHRU, 0xffff_0000, 8, 0x00ff_ff00);
+}
```

Comment:
maybe test here that the shift amount is indeed taken modulo the number of bits in the type?